### PR TITLE
Adjust Action Destination: First implementation + unit tests.

### PR DESCRIPTION
**Pull request by seg-leonelsanches**
One of our customers has dozens of apps, and hundreds (to not say thousands) of events that they want to send to Adjust. Additionally, these are all cloud events, sent by a service in their AWS environment. 

To solve the problem, I had to implement something more dynamic: considering a free-form mapping, the customer can send their App Token + Event Token as part of the event, if they wish to do it (and not only in the `context` section, or `integration` section, as the legacy integration works). 

This should work normally with the plugins and libraries that already exist.

## Testing

- [X] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [ ] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

### CODE REVIEW

**Comment by joe-ayoub-segment** [comment]
@smultani FYI - @seg-leonelsanches raised a PR for a new Destination for Adjust.

**Comment by joe-ayoub-segment** [comment]
hi @seg-leonelsanches thanks for raising this PR. 
I've asked my team how they'd like me to proceed with this review. Adjust is an important Destination which may need to be designed in such a way that it can be used by all customers. 

As a next step could we meet 1:1 to talk through the design and code please?

**Comment by seg-leonelsanches** [comment]
Hi @joe-ayoub-segment. Thanks for helping me with this. 

Yes, we can meet. Do you have my calendar?

**Comment by joe-ayoub-segment** [comment]
Hi @seg-leonelsanches - here's my Calendly. https://calendly.com/joe_ayoub/ 
Could we meet later in the week?

File: packages/destination-actions/src/destinations/adjust/index.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
+import type { DestinationDefinition } from '@segment/actions-core'
+import type { Settings } from './generated-types'
+
+import sendEvent from './sendEvent'
+
+const destination: DestinationDefinition<Settings> = {
+  name: 'Adjust',
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/adjust/index.ts:7]:
```suggestion
  name: 'Adjust (Actions)',
```

File: packages/destination-actions/src/destinations/adjust/sendEvent/index.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    },
+    currency: {
+      label: 'Currency',
+      description: 'The currency of the revenue. Only set if revenue is also set.',
+      type: 'string',
+      required: false
+    }
+  },
+  perform: (request, data) => {
+    const eventData = (data as any).rawData as { [key: string]: unknown }
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/adjust/sendEvent/index.ts:73]: Let's remove this line.

File: packages/destination-actions/src/destinations/adjust/sendEvent/index.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      description: 'The revenue amount.',
+      type: 'number',
+      required: false
+    },
+    currency: {
+      label: 'Currency',
+      description: 'The currency of the revenue. Only set if revenue is also set.',
+      type: 'string',
+      required: false
+    }
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/adjust/sendEvent/index.ts:70]: Add in a timestamp field

```
 timestamp: {
  label: 'Timestamp', 
  description: 'Timestamp for when the event happened',
  type: 'datetime'
 }
```

File: packages/destination-actions/src/destinations/adjust/functions.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    event_token: String(payload.event_token || settings.default_event_token),
+    environment: settings.environment,
+    s2s: 1,
+    callback_params: JSON.stringify(payload)
+  }
+
+  if (settings.send_event_creation_time) {
+    if (!eventData.timestamp) {
+      throw new Error('Event timestamp is required when send_event_creation_time is enabled.')
+    }
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/adjust/functions.ts:32]:
```suggestion
  if (settings.send_event_creation_time && !event.timestamp) {
      throw new Error('Event timestamp is required when send_event_creation_time is enabled.')
   }
```

File: packages/destination-actions/src/destinations/adjust/functions.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    adjustPayload.created_at_unix = parseInt((new Date(String(eventData.timestamp)).getTime() / 1000).toFixed(0))
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/adjust/functions.ts:34]:
```suggestion
    adjustPayload.created_at_unix = parseInt((new Date(String(payload.timestamp)).getTime() / 1000).toFixed(0))
```

File: packages/destination-actions/src/destinations/adjust/functions.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    const response = await request('https://s2s.adjust.com/event', {
+      method: 'POST',
+      body: JSON.stringify(event)
+    })
+
+    responses.push(response)
+  }
+
+  return responses
+}
```
Comment (joe-ayoub-segment) [inline packages/destination-actions/src/destinations/adjust/functions.ts:58]: Please use promise.all

**Review by joe-ayoub-segment — CHANGES_REQUESTED** [review]
Changes as discussed in meeting and inline.

**Comment by seg-leonelsanches** [comment]
@joe-ayoub-segment I think the destination is once again ready for review.

**Comment by joe-ayoub-segment** [comment]
Made some minor changes to get this ready for deploy.

**Review by joe-ayoub-segment — APPROVED** [review]

**Comment by joe-ayoub-segment** [comment]
hi @seg-leonelsanches PR deployed!

**Pull request merged**